plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/LaunchPipelineShortcut.java

```diff
@@ -0,0 +1,254 @@
+    ILaunchConfiguration[] dfLaunchConfigurations = getDataflowLaunchConfigurations();
+    String launchConfigurationName = resource.getLaunchName();
+    for (ILaunchConfiguration configuration : dfLaunchConfigurations) {
+      if (configuration.getName().equals(launchConfigurationName)) {
+        return new ILaunchConfiguration[] {configuration};
+      }
+    }
+    return new ILaunchConfiguration[0];
+  }
+
+  private LaunchableResource toLaunchableResource(IResource resource) {
+    if (resource == null) {
+      return null;
+    }
+    IJavaElement javaElement = (IJavaElement) resource.getAdapter(IJavaElement.class);
+    if (javaElement != null && javaElement.exists() && javaElement instanceof ICompilationUnit) {
+      ICompilationUnit compilationUnit = (ICompilationUnit) javaElement;
+      IType javaType = compilationUnit.findPrimaryType();
+      if (javaType == null) {
+        return null;
+      }
+      IMethod mainMethod = javaType.getMethod(
+          "main", new String[] {Signature.createTypeSignature("String[]", false)});
+      return new LaunchableResource(resource, mainMethod, javaType);
+    }
+    return new LaunchableResource(resource);
+  }
+
+  @Override
+  public IResource getLaunchableResource(ISelection selection) {
+    if (!selection.isEmpty() && selection instanceof IStructuredSelection) {
+      IStructuredSelection structuredSelection = (IStructuredSelection) selection;
+      Object element = structuredSelection.getFirstElement();
+      if (element instanceof IAdaptable) {
+        IAdaptable adaptable = (IAdaptable) element;
+        return (IResource) adaptable.getAdapter(IResource.class);
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public IResource getLaunchableResource(IEditorPart editorPart) {
+    return (IResource) editorPart.getEditorInput().getAdapter(IResource.class);
+  }
+
+  private ILaunchManager getLaunchManager() {
+    return DebugPlugin.getDefault().getLaunchManager();
+  }
+
+  /**
+   * A resource this shortcut is capable of launching. If the resource is a Java class file that
+   * contains a main method, contains a handle to that method and the primary type.
+   */
+  private static class LaunchableResource {
+    private final IResource resource;
+    /** The main method of the launchable resource. May be null. */
+    private final IMethod mainMethod;
+    /** The primary type of the launchable resource. Null if and only if mainMethod is null. */
+    private final IType primaryType;
+
+    public LaunchableResource(IResource resource) {
+      this(resource, null, null);
+    }
+
+    public LaunchableResource(IResource resource, IMethod mainMethod, IType primaryType) {
+      this.resource = resource;
+      if (mainMethod == null && primaryType != null) {
+        throw new NullPointerException(
+            "mainMethod and primaryType must either both be null or neither can be null");
```

Comment:
There should be two different exception messages here and below.This one could be

mainMethod is null but primaryType is not null